backend/cmd/worker/sample_validation.go

```diff
@@ -0,0 +1,285 @@
+	var visit func(*SampleValidationRecord)
+	visit = func(r *SampleValidationRecord) {
+		id := makeKey(r)
+		if visited[id] {
+			return
+		}
+		visited[id] = true
+		if r.Sample.SubmitterParentSampleId != "" {
+			parentID := fmt.Sprintf("%s:%s", r.Sample.SampleOrganizationCode, r.Sample.SubmitterParentSampleId)
+			if parent, ok := keyToRecord[parentID]; ok {
+				visit(parent)
+			}
+		}
+		sorted = append(sorted, r)
+	}
+
+	for _, r := range records {
+		visit(r)
+	}
+	copy(records, sorted)
+}
+
+func validateSamplesBatch(samples []types.SampleBatch, repoOrganization repository.OrganizationDAO, repoPatient repository.PatientsDAO, repoSample repository.SamplesDAO) ([]*SampleValidationRecord, error) {
+	records := make([]*SampleValidationRecord, 0, len(samples))
+	samplesMap := samplesMap(samples)
+
+	for index, sample := range samples {
+		record := &SampleValidationRecord{
+			BaseValidationRecord: BaseValidationRecord{Index: index},
+			Sample:               sample,
+		}
+
+		// 1. Validate fields
+		record.validateFieldLength("submitter_parent_sample_id", sample.SubmitterParentSampleId.String())
+		record.validateFieldLength("tissue_site", sample.TissueSite.String())
+
+		// 2. Validate patient
+		patient, patientErr := repoPatient.GetPatientByOrgCodeAndSubmitterPatientId(sample.PatientOrganizationCode, sample.SubmitterPatientId.String())
+		if patientErr != nil {
+			return nil, fmt.Errorf("error getting existing patient: %v", patientErr)
+		}
+		record.validatePatient(patient)
+
+		// 3. Validate organization
+		organization, orgErr := repoOrganization.GetOrganizationByCode(sample.SampleOrganizationCode)
+		if orgErr != nil {
+			return nil, fmt.Errorf("error getting existing sample organization: %v", orgErr)
+		}
+		record.validateOrganization(organization)
+
+		// 4. Validate if sample exists in DB
+		if organization != nil {
+			existingSample, sampleErr := repoSample.GetSampleBySubmitterSampleId(organization.ID, sample.SubmitterSampleId.String())
+			if sampleErr != nil {
+				return nil, fmt.Errorf("error getting existing sample: %v", sampleErr)
+			} else if existingSample != nil {
+				// 5. If exists, check if all fields are identical, and add error messages
+				record.validateExistingSampleInDb(existingSample)
+			}
+
+			// 6. Validate parent sample in DB if provided
+			if sample.SubmitterParentSampleId != "" {
+				existingParentSample, parentSampleErr := repoSample.GetSampleBySubmitterSampleId(organization.ID, sample.SubmitterParentSampleId.String())
+				if parentSampleErr != nil {
+					return nil, fmt.Errorf("error getting existing parent sample: %v", parentSampleErr)
+				}
+				record.validateExistingParentSampleInDb(existingParentSample)
+
+				if existingParentSample == nil {
+					// 7. If parent sample does not exist in DB, check if it exists in the current batch
+					_, parentSampleIsInBatch := samplesMap[fmt.Sprintf("%s:%s", sample.SampleOrganizationCode, sample.SubmitterParentSampleId)]
```

Comment:
use same function for building the map. Or you can use. strcut SampleKey {SampleOrganizationCode, SubmitterParentSampleId} as a key